node_modules/gh-auth/lib/api.js

```diff
@@ -153,8 +182,113 @@ var _validateLocalUserCredentials = function(app, email, password, done) {
+ * *themselves* but not any request in particular.
+ *
+ * As of now, the only request that they can do is exchange for an OAuth access token with an
+ * authorization code that has been authorized to it by a user. Theoretically, one could allow a
+ * client to update some metadata about itself, for example.
+ *
+ * @api private
+ */
+var _registerBasicClientPasswordAuthentication = function() {
+    passport.use(new BasicStrategy({'passReqToCallback': true}, _handleOauthClientAuthentication));
+};
+
+/**
+ * This is a permutation of `_registerBasicClientPasswordAuthentication` that allows a client to
+ * provide its credentials in the request body instead of basic authentication.
+ *
+ * For clients performing POST requests, this form of authentication is acceptable since there is
+ * CSRF protection in place to ensure that the request wasn't forged. However, for GET requests, the
+ * request should use `_BasicClientPasswordAuthentication` so that credentials are trasmitted in
+ * HTTP headers rather than the query string. This is a security measure as query strings should be
+ * safe to log in an access log, and client secrets are not safe for this.
+ *
+ * @api private
+ */
+var _registerClientPasswordAuthentication = function() {
+    passport.use(new ClientPasswordStrategy({'passReqToCallback': true}, _handleOauthClientAuthentication));
+};
+
+/**
+ * Convenience function for handling oauth client authentication, in which the client provides its
+ * id as the username and its secret as a password.
+ *
+ * @param  {Request}            req                 The request that initiated the authentication
+ * @param  {Number}             clientId            The id of the client trying to authenticate
+ * @param  {String}             clientSecret        The provided secret with which to authenticate
+ * @param  {Function}           done                Passport authentication `done` callback
+ * @param  {Object}             done.err            An error that occurred, if any
+ * @param  {Object|Boolean}     done.ghAuthInfo     If an object, should be the auth info object to apply to the request. Otherwise, should be `false` to indicate a failed authentcation
+ * @oaram  {Object}             done.errInfo        Additional error information, used by passport to assemble its errors
+ * @api private
+ */
+var _handleOauthClientAuthentication = function(req, clientId, clientSecret, done) {
+    clientId = GrassHopperUtil.getNumberParam(clientId);
+
+    var validator = new Validator();
+    validator.check(clientId, {'code': 400, 'msg': 'A valid client id must be specified'}).isInt();
+    validator.check(clientSecret, {'code': 400, 'msg': 'A valid client secret must be specified'}).notEmpty();
+    if (validator.hasErrors()) {
+        return done(null, false, {'message': validator.getFirstError().msg});
+    }
+
+    OauthDAO.getClient(clientId, function(err, client) {
+        if (err && err.code === 500) {
```

Comment:
For passport, `Internal Server Error` is the "first parameter is an err object" scenario, whereas other scenarios are an "authentication failed", in which case we use `false` as the second object.

If we put a 400 error or the like as the first parameter, passport would wrap it in a 500 which doesn't really make sense.